Adopt new NodeName enumeration more broadly in SVG code

cdumez opened the pull request:
#### d0ed0824740bddbcd4a112de942c8f7a256018ff
<pre>
Adopt new NodeName enumeration more broadly in SVG code
<a href="https://bugs.webkit.org/show_bug.cgi?id=255494">https://bugs.webkit.org/show_bug.cgi?id=255494</a>

Reviewed by Darin Adler.

* Source/WebCore/svg/SVGFEColorMatrixElement.cpp:
(WebCore::SVGFEColorMatrixElement::svgAttributeChanged):
* Source/WebCore/svg/SVGFEComponentTransferElement.cpp:
(WebCore::SVGFEComponentTransferElement::setFilterEffectAttributeFromChild):
* Source/WebCore/svg/SVGFECompositeElement.cpp:
(SVGFECompositeElement::setFilterEffectAttribute):
(SVGFECompositeElement::svgAttributeChanged):
(SVGFECompositeElement::createFilterEffect const): Deleted.
* Source/WebCore/svg/SVGFEConvolveMatrixElement.cpp:
(WebCore::SVGFEConvolveMatrixElement::setFilterEffectAttribute):
(WebCore::SVGFEConvolveMatrixElement::svgAttributeChanged):
* Source/WebCore/svg/SVGFEDiffuseLightingElement.cpp:
(WebCore::SVGFEDiffuseLightingElement::setFilterEffectAttribute):
(WebCore::SVGFEDiffuseLightingElement::svgAttributeChanged):
* Source/WebCore/svg/SVGFEDisplacementMapElement.cpp:
(WebCore::SVGFEDisplacementMapElement::setFilterEffectAttribute):
(WebCore::SVGFEDisplacementMapElement::svgAttributeChanged):
* Source/WebCore/svg/SVGFEDropShadowElement.cpp:
(WebCore::SVGFEDropShadowElement::svgAttributeChanged):
(WebCore::SVGFEDropShadowElement::setFilterEffectAttribute):
* Source/WebCore/svg/SVGFEGaussianBlurElement.cpp:
(WebCore::SVGFEGaussianBlurElement::svgAttributeChanged):
* Source/WebCore/svg/SVGFEMorphologyElement.cpp:
(SVGFEMorphologyElement::svgAttributeChanged):
* Source/WebCore/svg/SVGFEOffsetElement.cpp:
(WebCore::SVGFEOffsetElement::svgAttributeChanged):
* Source/WebCore/svg/SVGFESpecularLightingElement.cpp:
(WebCore::SVGFESpecularLightingElement::setFilterEffectAttribute):
* Source/WebCore/svg/SVGFETurbulenceElement.cpp:
(WebCore::SVGFETurbulenceElement::setFilterEffectAttribute):
* Source/WebCore/svg/SVGTests.cpp:
(WebCore::SVGTests::parseAttribute):
* Source/WebCore/svg/animation/SVGSMILElement.cpp:
(WebCore::SVGSMILElement::svgAttributeChanged):

Canonical link: <a href="https://commits.webkit.org/263011@main">https://commits.webkit.org/263011@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/142ea3abeb906546e39fe6157cc5699fe48e33cc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3305 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3364 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3480 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4722 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3636 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/3282 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3427 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3387 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2870 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3345 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3633 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2980 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/4544 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1136 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2945 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/2862 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2918 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2999 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/4281 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3378 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/2701 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2940 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2946 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/809 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2942 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3215 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/svg/SVGFEDiffuseLightingElement.cpp:
```diff
@@ -80,63 +80,67 @@ void SVGFEDiffuseLightingElement::attributeChanged(const QualifiedName& name, co
 {
     auto& feDiffuseLighting = downcast<FEDiffuseLighting>(effect);
+    auto& lightSource = feDiffuseLighting.lightSource().get();
+    const SVGFELightElement* lightElement = SVGFELightElement::findLightElement(this);
```
Since not al the cases use the light element, and this patch makes them all find it even if they don’t use it, I suggest using a lambda, and if can return a reference instead of a pointer. Maybe for `lightSource` too. Unless the idiom does not compile to some thing efficient.
```
lightSource().setAzimuth(element().azimuth());
```

Also not sure why the writer of this code felt they needed to put renderer into a local variable and asset i’s non-null. Null pointer dereferences aren’t that significantly harder to debug than assertion failures and a null renderer her does not seem like a practical problem; in fact a stale renderer seems just as likely and doesn’t get caught by an assert. So it should just be:
```
auto& style = renderer()->style();
```

darinadler commented on Source/WebCore/svg/SVGFEDropShadowElement.cpp:
```diff
@@ -85,51 +85,53 @@ void SVGFEDropShadowElement::attributeChanged(const QualifiedName& name, const A
         return feDropShadow.setShadowOpacity(style.svgStyle().floodOpacity());
-
+    default:
+        break;
```
I would have put the `ASSERT_NOT_REACHED` and the `return` in here.

cdumez replied:
My understanding was GCC didn't like this and would complain that the function is missing a return statement if I moved the return statement to the `default:` case.

darinadler replied:
I would like to confirm that at some point. I know one of the compilers does have an issue like this, but I think it’s about cases where all switch values are covered by cases and there is no default, not switch statements that have return statements in every case including a default.

darinadler commented on Source/WebCore/svg/SVGFEDropShadowElement.cpp:
```diff
@@ -85,51 +85,53 @@ void SVGFEDropShadowElement::attributeChanged(const QualifiedName& name, const A
-    if (attrName == SVGNames::flood_colorAttr)
+    switch (attrName.nodeName()) {
+    case AttributeNames::stdDeviationAttr:
+        return feDropShadow.setStdDeviationX(stdDeviationX()) || feDropShadow.setStdDeviationY(stdDeviationY());
```
Here we fetch style even when we don’t use it. But we didn’t do that before the patch. I suggest moving the style fetching into the cases even though that requires braces for `flood_color`. Again without the null pointer assertion.

darinadler commented on Source/WebCore/svg/SVGFESpecularLightingElement.cpp:
```diff
@@ -86,46 +86,45 @@ void SVGFESpecularLightingElement::attributeChanged(const QualifiedName& name, c
 bool SVGFESpecularLightingElement::setFilterEffectAttribute(FilterEffect& effect, const QualifiedName& attrName)
 {
     auto& feSpecularLighting = downcast<FESpecularLighting>(effect);
+    auto& lightSource = feSpecularLighting.lightSource().get();
```
Same issues here with `lightElement` and `style` that I mentioned above. Seems unfortunate to compute and not use them for some cases.

darinadler commented on Source/WebCore/svg/SVGFEComponentTransferElement.cpp:
```diff
@@ -106,27 +106,24 @@ bool SVGFEComponentTransferElement::setFilterEffectAttributeFromChild(FilterEffe
 
-    if (attrName == SVGNames::typeAttr)
+    switch (attrName.nodeName()) {
+    case AttributeNames::typeAttr:
```
Seems to me these functions are extra wordy because of the long local variable names. I would rename the arguments to `filterEffect` and then use `effect` for the downcast result local instead of unwieldy long names like `feComponentTransfer`.

darinadler reviewed (approved):
Just fine as is. I have a few ideas, but please consider them all optional.

cdumez committed:
Adopt new NodeName enumeration more broadly in SVG code
https://bugs.webkit.org/show_bug.cgi?id=255494

Reviewed by Darin Adler.

* Source/WebCore/svg/SVGFEColorMatrixElement.cpp:
(WebCore::SVGFEColorMatrixElement::svgAttributeChanged):
* Source/WebCore/svg/SVGFEComponentTransferElement.cpp:
(WebCore::SVGFEComponentTransferElement::setFilterEffectAttributeFromChild):
* Source/WebCore/svg/SVGFECompositeElement.cpp:
(SVGFECompositeElement::setFilterEffectAttribute):
(SVGFECompositeElement::svgAttributeChanged):
(SVGFECompositeElement::createFilterEffect const): Deleted.
* Source/WebCore/svg/SVGFEConvolveMatrixElement.cpp:
(WebCore::SVGFEConvolveMatrixElement::setFilterEffectAttribute):
(WebCore::SVGFEConvolveMatrixElement::svgAttributeChanged):
* Source/WebCore/svg/SVGFEDiffuseLightingElement.cpp:
(WebCore::SVGFEDiffuseLightingElement::setFilterEffectAttribute):
(WebCore::SVGFEDiffuseLightingElement::svgAttributeChanged):
* Source/WebCore/svg/SVGFEDisplacementMapElement.cpp:
(WebCore::SVGFEDisplacementMapElement::setFilterEffectAttribute):
(WebCore::SVGFEDisplacementMapElement::svgAttributeChanged):
* Source/WebCore/svg/SVGFEDropShadowElement.cpp:
(WebCore::SVGFEDropShadowElement::svgAttributeChanged):
(WebCore::SVGFEDropShadowElement::setFilterEffectAttribute):
* Source/WebCore/svg/SVGFEGaussianBlurElement.cpp:
(WebCore::SVGFEGaussianBlurElement::svgAttributeChanged):
* Source/WebCore/svg/SVGFEMorphologyElement.cpp:
(SVGFEMorphologyElement::svgAttributeChanged):
* Source/WebCore/svg/SVGFEOffsetElement.cpp:
(WebCore::SVGFEOffsetElement::svgAttributeChanged):
* Source/WebCore/svg/SVGFESpecularLightingElement.cpp:
(WebCore::SVGFESpecularLightingElement::setFilterEffectAttribute):
* Source/WebCore/svg/SVGFETurbulenceElement.cpp:
(WebCore::SVGFETurbulenceElement::setFilterEffectAttribute):
* Source/WebCore/svg/SVGTests.cpp:
(WebCore::SVGTests::parseAttribute):
* Source/WebCore/svg/animation/SVGSMILElement.cpp:
(WebCore::SVGSMILElement::svgAttributeChanged):

Canonical link: https://commits.webkit.org/263011@main
